### PR TITLE
runtime: Fix ordering of trace spans

### LINKED ISSUE #1355: runtime: some runtime trace spans have incorrect parent/child relationship
Related to issue #1347 

Some runtime trace spans have incorrect parent/child ordering which in some cases can be fixed by returning the context when creating a span:
```
span, ctx := opentracing.StartSpanFromContext(ctx, name)
```

Another, `bindMount`, appears parentless in the output because in [kata_agent.go](https://github.com/kata-containers/kata-containers/blob/main/src/runtime/virtcontainers/kata_agent.go#L439) the background context is being passed to `bindMount()`. It can be switched to `k.ctx` although then this particular span does not show up in the trace output (either parentless or grouped under `New()`).

### PR DESCRIPTION
**Pull request by cmaf**
Change whether trace function returns context in some functions in
virtcontainers to fix the ordering of trace spans. Use existing context
rather than background context in Run() so that span exists as a child
of a parent span.

Fixes #1355

Signed-off-by: Chelsea Mafrica <chelsea.e.mafrica@intel.com>

### CODE REVIEW

**Comment by liubin** [comment]
@cmaf thanks for this PR.

But use a global context ( `s.ctx`) may lead span parent/child relationship incorrect.

The spans created:

image

The source:
image

Maybe in some cases passing context as parameters explicitly is a good option.

**Comment by cmaf** [comment]
@liubin thanks! I noticed this and added the WIP yesterday, fixing it now.

**Review by jodh-intel — APPROVED** [review]
Thanks @cmaf - I haven't actually tested, but lgtm

/cc @liubin.

**Comment by cmaf** [comment]
@liubin this is updated, would you be able to test locally?

There will be some isolated `sendReq()` spans which I created a separate issue for. https://github.com/kata-containers/kata-containers/issues/1504

File: src/runtime/virtcontainers/clh.go [hunk 1]
```diff
@@ -473,8 +474,8 @@ func (clh *cloudHypervisor) hotPlugVFIODevice(device config.VFIODev) error {
 	return err
 }
 
-func (clh *cloudHypervisor) hotplugAddDevice(devInfo interface{}, devType deviceType) (interface{}, error) {
-	span, _ := clh.trace("hotplugAddDevice")
+func (clh *cloudHypervisor) hotplugAddDevice(ctx context.Context, devInfo interface{}, devType deviceType) (interface{}, error) {
+	span, ctx := clh.trace(ctx, "hotplugAddDevice")
```
Comment (liubin) [inline src/runtime/virtcontainers/clh.go:478]: `ctx` declared but not used?

**Comment by liubin** [comment]
Thank you @cmaf , LGTM.  

Only one point should be fixed I think.

`CreateContainer` will create two separated traces:

- `kata: create`
- `kata: createContainer`

besides `kata: root span`.

And this patch is one that can fix it without updating `s.ctx`:

```diff
diff --git a/src/runtime/containerd-shim-v2/create.go b/src/runtime/containerd-shim-v2/create.go
index 58491f43..8619480c 100644
--- a/src/runtime/containerd-shim-v2/create.go
+++ b/src/runtime/containerd-shim-v2/create.go
@@ -116,7 +116,7 @@ func create(ctx context.Context, s *service, r *taskAPI.CreateTaskRequest) (*con
 
        case vc.PodContainer:
                var span otelTrace.Span
-               span, s.ctx = trace(ctx, "create")
+               span, ctx = trace(s.ctx, "create")
                defer span.End()
 
                if s.sandbox == nil {
```

`katautils.CreateContainer()` may need context that contains parent span info.

**Review by liubin — APPROVED** [review]

**Comment by cmaf** [comment]
thanks @liubin, I made these changes. 

/test

**Pull request merged**